jiant/preprocess.py

```diff
@@ -743,53 +743,171 @@ class ModelPreprocessingInterface(object):
+            *args, trunc_strategy=None, trunc_side=None, **kwargs
+        ):
+            """
+            Calls the apply_boundary_tokens function provided in the parent function and, if the
+            output exceeds the max number of tokens, applies a truncation strategy to the inputs,
+            then re-applies the apply_boundary_tokens function on the truncated inputs and returns
+            the result.
+
+            Parameters
+            ----------
+            args : tuple
+                see args docs for apply_boundary_tokens and apply_lm_boundary_tokens in
+                huggingface_transformers_interface.modules
+            trunc_strategy : str
+                Strategy (e.g., default "trunc_s2", or optionally "trunc_s1") determining which
+                input to trunc to reduce the total number of tokens returned to within max_tokens.
+            trunc_side : str
+                Which side of the sequence to truncate: "left" or "right".
+            kwargs : dict
+                see args docs for apply_boundary_tokens and apply_lm_boundary_tokens in
+                huggingface_transformers_interface.modules
+
+            Returns
+            -------
+            seq_w_boundry_tokens : List[str]
+                List of tokens returned by applying the specified apply_boundary_tokens function
+                to the inputs and using the specified truncation strategy.
+
+            """
+            seq_w_boundry_tokens = apply_boundary_tokens(*args, **kwargs)
+            # if after calling the apply_boundary_tokens function the number of tokens is greater
+            # than the model's max, a truncation strategy can reduce the number of tokens:
+            num_excess_tokens = len(seq_w_boundry_tokens) - max_tokens
+            if num_excess_tokens > 0:
+                if not (trunc_strategy and trunc_side):
+                    raise ValueError(
+                        "Input(s) will exceed model capacity or max_seq_len. "
+                        + "Adjust settings as necessary, or to automatically truncate "
+                        + "inputs in apply_boundary_tokens, call apply_boundary_tokens"
+                        + "with trunc_strategy and trunc_side keyword args."
+                    )
+                log.warning(
+                    "After applying boundry tokens, sequence length for this example is "
+                    + str(len(seq_w_boundry_tokens))
+                    + ", max_seq_len is "
+                    + str(max_tokens)
+                    + ". Truncation strategy "
+                    + trunc_strategy
+                    + " will be applied, removing elements from the "
+                    + trunc_side
+                    + " side of the input."
+                )
+                if trunc_strategy == "trunc_s2":
+                    s2 = kwargs["s2"]
+                    log.warning("Before truncation, s2 length = " + str(len(s2)))
```

Comment:
incredibly inconsequential: I'd prefer to just print the before/after length on the same line. Some datasets might have lots of lines to be truncated.

Reply:
Sounds like a good idea to me — done in 09cb503.